File: examples/library-examples/test/Support/cache-purge.ts
/** Custom-capability fixture, never an assertion of native Cloudflare support. */
export async function exerciseCachePurge(
  invoke: (context: object, operation: string) => Promise<string>,
  request: Request,
): Promise<Response> {
  const url = new URL(request.url);
  const scenario = url.searchParams.get("scenario") ?? "success";
  const operation = url.searchParams.get("operation") ?? "tags";
  const calls: unknown[] = [];
  const context =
    scenario === "missing"
      ? {}
      : {
          cache: {
            async purge(options: unknown): Promise<unknown> {
              calls.push(options);
              if (scenario === "throw") {
                throw new Error("injected custom purge capability failure");
              }
              if (scenario === "malformed-success") {
                return { success: "true", errors: [] };
              }
              if (scenario === "null") {
                return null;
              }
              if (scenario === "rejected") {
                return {
                  success: false,
                  errors: [{ code: 1001, message: "fixture policy refusal" }],
                };
              }
              return { success: true, errors: [] };
            },
          },
        };
  const outcome: unknown = JSON.parse(await invoke(context, operation));
  return Response.json({ outcome, calls });
}

File: examples/library-examples/src/LibraryExamples/CachePurge.hs
-- | Contract example for the library's custom ctx.cache.purge capability.
-- Standard Cloudflare ExecutionContext does not supply this capability.
module LibraryExamples.CachePurge (runCachePurge) where

import Cloudflare.Workers.Cache
import ExampleSupport.Interop (jsValToText, textToJSVal)
import Cloudflare.Workers.Reactor (WorkersExecutionContext(..))
import Control.Exception (try)
import Data.Aeson (Value, encode, object, (.=))
import Data.ByteString.Lazy qualified as Lazy
import Data.Text.Encoding (decodeUtf8)
import GHC.Wasm.Prim (JSVal)

-- | Choose one purge operation and preserve the structured provider outcome.
-- The caller supplies the explicit custom capability; no native support is assumed.
runCachePurge :: JSVal -> JSVal -> IO JSVal
runCachePurge contextValue modeValue = do
  mode <- jsValToText modeValue
  let options = case mode of
        "tags" -> Just (PurgeTags ["guide", "catalog"])
        "prefixes" -> Just (PurgePathPrefixes ["/guide/", "/catalog/"])
        "everything" -> Just PurgeEverything
        _ -> Nothing
  value <- case options of
    Nothing -> pure (object ["outcome" .= ("invalid-operation" :: String)])
    Just selected -> do
      result <- try @CachePurgeFailed (cachePurge (WorkersExecutionContext contextValue) selected)
      pure $ case result of
        Left _ -> object ["outcome" .= ("capability-failed" :: String)]
        Right response -> object
          [ "outcome" .= ("resolved" :: String)
          , "success" .= cachePurgeResultSuccess response
          , "errors" .= fmap errorValue (cachePurgeResultErrors response)
          ]
  textToJSVal (decodeUtf8 (Lazy.toStrict (encode value)))
  where
    errorValue :: CachePurgeError -> Value
    errorValue failure = object
      [ "code" .= cachePurgeErrorCode failure
      , "message" .= cachePurgeErrorMessage failure
      ]
